sdk/src/wallet/operations/transaction/high_level/staking/end.rs

```diff
@@ -0,0 +1,76 @@
+// Copyright 2024 IOTA Stiftung
+// SPDX-License-Identifier: Apache-2.0
+
+use crate::{
+    client::{api::PreparedTransactionData, secret::SecretManage},
+    types::block::{
+        context_input::{ContextInput, RewardContextInput},
+        output::{AccountId, AccountOutputBuilder},
+    },
+    wallet::{types::TransactionWithMetadata, TransactionOptions, Wallet},
+};
+
+impl<S: 'static + SecretManage> Wallet<S>
+where
+    crate::wallet::Error: From<S::Error>,
+    crate::client::Error: From<S::Error>,
+{
+    pub async fn end_staking(&self, account_id: AccountId) -> crate::wallet::Result<TransactionWithMetadata> {
+        let prepared = self.prepare_end_staking(account_id).await?;
+
+        self.sign_and_submit_transaction(prepared, None, None).await
+    }
+
+    /// Prepares the transaction for [Wallet::end_staking()].
+    pub async fn prepare_end_staking(&self, account_id: AccountId) -> crate::wallet::Result<PreparedTransactionData> {
+        log::debug!("[TRANSACTION] prepare_end_staking");
+
+        let account_output_data = self
+            .data()
+            .await
+            .unspent_account_output(&account_id)
+            .cloned()
+            .ok_or_else(|| crate::wallet::Error::AccountNotFound)?;
+
+        let staking_feature = account_output_data
+            .output
+            .features()
+            .and_then(|f| f.staking())
+            .ok_or_else(|| crate::wallet::Error::StakingFailed(format!("account id {account_id} is not staking")))?;
+
+        let protocol_parameters = self.client().get_protocol_parameters().await?;
+
+        let slot_commitment_id = self.client().get_issuance().await?.latest_commitment.id();
+        let future_bounded_epoch =
+            protocol_parameters.epoch_index_of(protocol_parameters.future_bounded_slot(slot_commitment_id));
+
+        if future_bounded_epoch <= staking_feature.end_epoch() {
+            let end_epoch = protocol_parameters.epoch_index_of(slot_commitment_id.slot_index())
+                + (staking_feature.end_epoch() - future_bounded_epoch);
```

Comment:
I have a hard time wrapping my head around that, why not just log `staking_feature.end_epoch`?

Reply:
I figured since the future bounded epoch is the deciding factor, then it had to be relative to that value